cuda/btx_cudainterval_callbacks.cpp

```diff
@@ -0,0 +1,589 @@
+                                        int64_t ts, const char *event_class_name,
+                                        const char *hostname, int64_t vpid,
+                                        uint64_t vtid, CUdevice dev) {
+  auto state = static_cast<data_t *>(usr_data);
+  state->context_manager.primary_ctx_release_entry({hostname, vpid, vtid}, dev);
+}
+
+void primary_ctx_release_exit_callback(void *btx_handle, void *usr_data,
+                                       int64_t ts, const char *event_class_name,
+                                       const char *hostname, int64_t vpid,
+                                       uint64_t vtid, CUresult cuResult) {
+  auto state = static_cast<data_t *>(usr_data);
+  state->context_manager.primary_ctx_release_exit({hostname, vpid, vtid}, cuResult);
+}
+
+void primary_ctx_reset_entry_callback(void *btx_handle, void *usr_data,
+                                      int64_t ts, const char *event_class_name,
+                                      const char *hostname, int64_t vpid,
+                                      uint64_t vtid, CUdevice dev) {
+  auto state = static_cast<data_t *>(usr_data);
+  state->context_manager.primary_ctx_reset_entry({hostname, vpid, vtid}, dev);
+}
+
+void primary_ctx_reset_exit_callback(void *btx_handle, void *usr_data,
+                                     int64_t ts, const char *event_class_name,
+                                     const char *hostname, int64_t vpid,
+                                     uint64_t vtid, CUresult cuResult) {
+  auto state = static_cast<data_t *>(usr_data);
+  state->context_manager.primary_ctx_reset_exit({hostname, vpid, vtid}, cuResult);
+}
+
+void ctx_create_entry_callback(void *btx_handle, void *usr_data,
+                               int64_t ts, const char *event_class_name,
+                               const char *hostname, int64_t vpid,
+                               uint64_t vtid, CUdevice dev) {
+  auto state = static_cast<data_t *>(usr_data);
+  state->context_manager.ctx_create_entry({hostname, vpid, vtid}, dev);
+}
+
+void ctx_create_exit_callback(void *btx_handle, void *usr_data,
+                              int64_t ts, const char *event_class_name,
+                              const char *hostname, int64_t vpid,
+                              uint64_t vtid, CUresult cuResult,
+                              CUcontext ctx) {
+  auto state = static_cast<data_t *>(usr_data);
+  state->context_manager.ctx_create_exit({hostname, vpid, vtid}, cuResult, ctx);
+}
+
+void ctx_destroy_entry_callback(void *btx_handle, void *usr_data,
+                                int64_t ts, const char *event_class_name,
+                                const char *hostname, int64_t vpid,
+                                uint64_t vtid, CUcontext ctx) {
+  auto state = static_cast<data_t *>(usr_data);
+  state->context_manager.ctx_destroy_entry({hostname, vpid, vtid}, ctx);
+}
```

Comment:
Should we replace all those stuff by some MACRO? If the name match should be doable.

Reply:
I don't love macros for this. It's ugly but clear now? Are you thinking something like this:

`#define STATE_CAST(data)  static_cast<data_t *>(data)`

and

`auto state = STATE_CAST(usr_data)`

Not totally sure macros even work for static cast, I guess it should.

Reply:
No sorry was thinkinking for about 

```
DEFINE CONTEXT_MANAGER_CALLBACK_FOO


void $FOO_callback(...) {
   state->context_manage.$FOO(...)
}
```